Use notification center delegate events to drive close and click events

beidson opened the pull request:
#### c22b64a0bc64c82042e780557fde3f38bbefb04e
<pre>
Use notification center delegate events to drive close and click events
<a href="https://rdar.apple.com/134990333">rdar://134990333</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=278897">https://bugs.webkit.org/show_bug.cgi?id=278897</a>

Reviewed by Ben Nham.

Relying on the other mechanism for this required more work than is possible at this time.

This patch adds:
- SPI for clients to pass in UNNotificationResponse objects to be handled
- Changes the default notification center delegate to use that SPI for automatic handling
  in cases where the client doesn&apos;t have their own notification center delegate

* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[_WKWebsiteDataStoreNotificationCenterDelegate userNotificationCenter:didReceiveNotificationResponse:withCompletionHandler:]):
(+[WKWebsiteDataStore handleNotificationResponse:]):
(-[_WKWebsiteDataStoreBSActionHandler handleNotificationResponse:]):
(-[_WKWebsiteDataStoreBSActionHandler _respondToApplicationActions:fromTransitionContext:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h:

Canonical link: <a href="https://commits.webkit.org/283013@main">https://commits.webkit.org/283013@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/98ad0751aeed6f88370923092625ca0717df0b3d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/64925 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44292 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/17539 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/68949 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/15531 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/52075 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/15813 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/52161 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/10715 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/67991 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/40955 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/56182 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32783 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/37626 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14407 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/59535 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/13892 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/70654 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/8876 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/13373 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/59489 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/8909 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56243 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/59693 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14332 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7318 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/989 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/40104 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/41181 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/42362 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/40925 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

bnham commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -362,17 +362,22 @@ void didExceedMemoryFootprintThreshold(size_t footprint, const String& domain, u
+
 - (void)userNotificationCenter:(UNUserNotificationCenter *)center didReceiveNotificationResponse:(UNNotificationResponse *)response withCompletionHandler:(void (^)())completionHandler
 {
+    [WKWebsiteDataStore handleNotificationResponse:response];
```
Something I remembered is that we have to have this (as well as the BSAction handler that fires the chain of push events) take a UIApp background task for the duration of the event firing. Or we have to provide a completion handler to the client so they can do the equivalent thing. Otherwise the app can get suspended in the middle of event processing.

beidson replied:
I'm filing a followup radar for this

bnham reviewed (approved):
LGTM but we need to fix the issue with background tasks in this patch or a subsequent patch (if subsequent then file a radar to track the fix).

beidson committed:
Use notification center delegate events to drive close and click events
rdar://134990333
https://bugs.webkit.org/show_bug.cgi?id=278897

Reviewed by Ben Nham.

Relying on the other mechanism for this required more work than is possible at this time.

This patch adds:
- SPI for clients to pass in UNNotificationResponse objects to be handled
- Changes the default notification center delegate to use that SPI for automatic handling
  in cases where the client doesn't have their own notification center delegate

* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[_WKWebsiteDataStoreNotificationCenterDelegate userNotificationCenter:didReceiveNotificationResponse:withCompletionHandler:]):
(+[WKWebsiteDataStore handleNotificationResponse:]):
(-[_WKWebsiteDataStoreBSActionHandler handleNotificationResponse:]):
(-[_WKWebsiteDataStoreBSActionHandler _respondToApplicationActions:fromTransitionContext:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h:

Canonical link: https://commits.webkit.org/283013@main
